diff --git a/chrome/browser/browser_resources.grd b/chrome/browser/browser_resources.grd
index 87c529b..6ac14788 100644
--- a/chrome/browser/browser_resources.grd
+++ b/chrome/browser/browser_resources.grd
@@ -87,6 +87,7 @@
       </if>
     </structures>
     <includes>
+      <include name="IDR_SODIUM_JS" file="resources\sodium.js" type="BINDATA" skip_minify="true" />
       <if expr="is_win or is_macosx or desktop_linux or chromeos">
         <include name="IDR_DISCARDS_MOJO_API_HTML" file="resources\discards\mojo_api.html" compress="gzip" allowexternalscript="true" type="BINDATA" />
         <include name="IDR_DISCARDS_DATABASE_TAB_HTML" file="resources\discards\database_tab.html" compress="gzip" allowexternalscript="true" type="BINDATA" />
diff --git a/chrome/renderer/chrome_content_renderer_client.cc b/chrome/renderer/chrome_content_renderer_client.cc
index 6ad2480..870eba5 100644
--- a/chrome/renderer/chrome_content_renderer_client.cc
+++ b/chrome/renderer/chrome_content_renderer_client.cc
@@ -40,6 +40,7 @@
 #include "chrome/common/thread_profiler.h"
 #include "chrome/common/url_constants.h"
 #include "chrome/common/webui_url_constants.h"
+#include "chrome/grit/browser_resources.h"
 #include "chrome/grit/chromium_strings.h"
 #include "chrome/grit/generated_resources.h"
 #include "chrome/grit/renderer_resources.h"
@@ -138,6 +139,7 @@
 #include "third_party/blink/public/web/web_origin_trials.h"
 #include "third_party/blink/public/web/web_plugin_container.h"
 #include "third_party/blink/public/web/web_plugin_params.h"
+#include "third_party/blink/public/web/web_script_source.h"
 #include "third_party/blink/public/web/web_security_policy.h"
 #include "third_party/blink/public/web/web_view.h"
 #include "ui/base/l10n/l10n_util.h"
@@ -145,6 +147,7 @@
 #include "ui/base/resource/resource_bundle.h"
 #include "ui/base/webui/jstemplate_builder.h"
 #include "url/origin.h"
+#include "url/sodium_whitelist.h"
 
 #if defined(OS_ANDROID)
 #include "chrome/renderer/sandbox_status_extension_android.h"
@@ -1458,6 +1461,21 @@ void ChromeContentRendererClient::RecordRapporURL(const std::string& metric,
   rappor_recorder_->RecordRapporURL(metric, url);
 }
 
+void InjectSodium(content::RenderFrame* render_frame) {
+  WebLocalFrame* frame = render_frame->GetWebFrame();
+  if (!frame) return;
+
+  GURL url = GURL(frame->GetDocument().Url());
+  LOG(INFO) << "frame.url: " << url.spec();
+  LOG(INFO) << "render_frame.main_frame: " << render_frame->IsMainFrame();
+  LOG(INFO) << "render_frame.enable_sodium: " << render_frame->IsEnableSodium();
+
+  if (render_frame->IsMainFrame() && render_frame->IsEnableSodium() && url::SodiumWhitelist::IsAllowSodiumJs(url)) {
+    base::StringPiece source_piece = ui::ResourceBundle::GetSharedInstance().GetRawDataResource(IDR_SODIUM_JS);
+    render_frame->ExecuteJavaScript(base::UTF8ToUTF16(source_piece));
+  }
+}
+
 void ChromeContentRendererClient::RunScriptsAtDocumentStart(
     content::RenderFrame* render_frame) {
 #if BUILDFLAG(ENABLE_EXTENSIONS)
@@ -1474,6 +1492,8 @@ void ChromeContentRendererClient::RunScriptsAtDocumentEnd(
       render_frame);
   // |render_frame| might be dead by now.
 #endif
+
+  InjectSodium(render_frame);
 }
 
 void ChromeContentRendererClient::RunScriptsAtDocumentIdle(
