diff --git a/chrome/android/chrome_java_sources.gni b/chrome/android/chrome_java_sources.gni
index 607431c..6536223 100644
--- a/chrome/android/chrome_java_sources.gni
+++ b/chrome/android/chrome_java_sources.gni
@@ -1372,6 +1372,7 @@ chrome_java_sources = [
   "java/src/org/chromium/chrome/browser/settings/NfcSystemLevelSetting.java",
   "java/src/org/chromium/chrome/browser/settings/SettingsActivity.java",
   "java/src/org/chromium/chrome/browser/settings/SettingsLauncher.java",
+  "java/src/org/chromium/chrome/browser/settings/SodiumSettings.java",
   "java/src/org/chromium/chrome/browser/settings/about/AboutChromePreferenceOSVersion.java",
   "java/src/org/chromium/chrome/browser/settings/about/AboutChromeSettings.java",
   "java/src/org/chromium/chrome/browser/settings/about/AboutSettingsBridge.java",
diff --git a/chrome/android/java/res/menu/main_menu.xml b/chrome/android/java/res/menu/main_menu.xml
index f2c7c73..7c63358 100644
--- a/chrome/android/java/res/menu/main_menu.xml
+++ b/chrome/android/java/res/menu/main_menu.xml
@@ -83,6 +83,16 @@
             android:icon="@drawable/reader_mode_prefs_icon" />
         <item android:id="@+id/preferences_id"
             android:title="@string/menu_settings" />
+        <item android:id="@+id/enable_sodium_row_menu_id"
+            android:title="@null">
+          <menu>
+              <item android:id="@+id/enable_sodium_id"
+                android:title="@string/sodium_title" />
+              <item android:id="@+id/enable_sodium_check_id"
+                android:title="@null"
+                android:checkable="true" />
+          </menu>
+        </item>
         <item android:id="@+id/help_id"
             android:title="@string/menu_help" />
         <item android:id="@+id/enter_vr_id"
diff --git a/chrome/android/java/res/xml/main_preferences.xml b/chrome/android/java/res/xml/main_preferences.xml
index 99288f0..d0499e00 100644
--- a/chrome/android/java/res/xml/main_preferences.xml
+++ b/chrome/android/java/res/xml/main_preferences.xml
@@ -66,44 +66,49 @@
         android:order="11"
         android:title="@string/prefs_section_advanced"/>
     <Preference
+        android:fragment="org.chromium.chrome.browser.settings.SodiumSettings"
+        android:key="sodium"
+        android:order="12"
+        android:title="@string/sodium_title"/>
+    <Preference
         android:fragment="org.chromium.chrome.browser.settings.privacy.PrivacySettings"
         android:key="privacy"
-        android:order="12"
+        android:order="13"
         android:title="@string/prefs_privacy"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.settings.accessibility.AccessibilitySettings"
         android:key="accessibility"
-        android:order="13"
+        android:order="14"
         android:title="@string/prefs_accessibility"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.settings.website.SiteSettings"
         android:key="content_settings"
-        android:order="14"
+        android:order="15"
         android:title="@string/prefs_site_settings"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.settings.languages.LanguageSettings"
         android:key="languages"
-        android:order="15"
+        android:order="16"
         android:title="@string/language_settings"/>
     <org.chromium.chrome.browser.settings.ChromeBasePreference
         android:fragment="org.chromium.chrome.browser.settings.datareduction.DataReductionPreferenceFragment"
         android:key="data_reduction"
-        android:order="16"
+        android:order="17"
         android:title="@string/data_reduction_title_lite_mode"/>
     <org.chromium.chrome.browser.settings.ChromeBasePreference
         android:fragment="org.chromium.chrome.browser.settings.download.DownloadSettings"
         android:key="downloads"
-        android:order="17"
+        android:order="18"
         android:title="@string/menu_downloads"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.settings.developer.DeveloperSettings"
         android:key="developer"
-        android:order="18"
+        android:order="19"
         android:title="Developer options"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.settings.about.AboutChromeSettings"
         android:key="about_chrome"
-        android:order="19"
+        android:order="20"
         android:title="@string/prefs_about_chrome"/>
 
 </PreferenceScreen>
diff --git a/chrome/android/java/res/xml/sodium_preferences.xml b/chrome/android/java/res/xml/sodium_preferences.xml
new file mode 100644
index 0000000..ae72e95
--- /dev/null
+++ b/chrome/android/java/res/xml/sodium_preferences.xml
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<PreferenceScreen
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <org.chromium.chrome.browser.settings.ChromeSwitchPreference
+        android:key="sodium_switch"
+        android:summaryOn="@string/text_on"
+        android:summaryOff="@string/text_off" />
+
+    <org.chromium.chrome.browser.settings.TextMessagePreference
+        android:key="sodium_description"
+        android:title="@string/sodium_description"
+        app:allowDividerBelow="false" />
+
+</PreferenceScreen>
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java b/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java
index 8201302..5552066 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java
@@ -2097,6 +2097,10 @@ public abstract class ChromeActivity<C extends ChromeActivityComponent>
             currentTab.getWebContents().getNavigationController().setUseDesktopUserAgent(
                     !usingDesktopUserAgent, reloadOnChange);
             RecordUserAction.record("MobileMenuRequestDesktopSite");
+        } else if (id == R.id.enable_sodium_id || id == R.id.enable_sodium_check_id) {
+            final boolean reloadOnChange = !currentTab.isNativePage();
+            final boolean enabled = currentTab.getEnableSodium();
+            currentTab.setEnableSodium(!enabled, reloadOnChange);
         } else if (id == R.id.reader_mode_prefs_id) {
             DomDistillerUIUtils.openSettings(currentTab.getWebContents());
         } else {
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java b/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java
index 79faee2..03973d0 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java
@@ -282,6 +282,7 @@ public class AppMenuPropertiesDelegateImpl implements AppMenuPropertiesDelegate
             prepareAddToHomescreenMenuItem(menu, currentTab, canShowHomeScreenMenuItem);
 
             updateRequestDesktopSiteMenuItem(menu, currentTab, true /* can show */);
+            updateEnableSodiumMenuItem(menu, currentTab, false /* temporary don't show */);
 
             // Only display reader mode settings menu option if the current page is in reader mode.
             menu.findItem(R.id.reader_mode_prefs_id)
@@ -456,6 +457,38 @@ public class AppMenuPropertiesDelegateImpl implements AppMenuPropertiesDelegate
     @Override
     public void onFooterViewInflated(AppMenuHandler appMenuHandler, View view) {}
 
+    /**
+     * @param menu {@link Menu} for enable sodium.
+     * @param currentTab      Current tab being displayed.
+     */
+    protected void updateEnableSodiumMenuItem(Menu menu, Tab currentTab, boolean canShowEnableSodium) {
+        MenuItem menuRow   = menu.findItem(R.id.enable_sodium_row_menu_id);
+        MenuItem menuLabel = menu.findItem(R.id.enable_sodium_id);
+        MenuItem menuCheck = menu.findItem(R.id.enable_sodium_check_id);
+
+        // Hide enable sodium on all chrome:// pages except for the NTP.
+        String url = currentTab.getUrl();
+        boolean isChromeScheme = url.startsWith(UrlConstants.CHROME_URL_PREFIX)
+                || url.startsWith(UrlConstants.CHROME_NATIVE_URL_PREFIX);
+        // Also hide enable sodium on Reader Mode.
+        boolean isDistilledPage = DomDistillerUrlUtils.isDistilledPage(url);
+
+        boolean itemVisible = canShowEnableSodium
+                && (!isChromeScheme || currentTab.isNativePage()) && !isDistilledPage;
+        menuRow.setVisible(itemVisible);
+        if (!itemVisible) return;
+
+        boolean enabled = currentTab.getEnableSodium();
+        // Mark the checkbox if sodium is activated on this page.
+        menuCheck.setChecked(enabled);
+
+        // This title doesn't seem to be displayed by Android, but it is used to set up
+        // accessibility text in {@link AppMenuAdapter#setupMenuButton}.
+        menuLabel.setTitleCondensed(enabled
+                        ? mContext.getString(R.string.sodium_on)
+                        : mContext.getString(R.string.sodium_off));
+    }
+
     @Override
     public void onHeaderViewInflated(AppMenuHandler appMenuHandler, View view) {}
 
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/settings/MainSettings.java b/chrome/android/java/src/org/chromium/chrome/browser/settings/MainSettings.java
index 0991b64..c0c787a 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/settings/MainSettings.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/settings/MainSettings.java
@@ -25,6 +25,8 @@ import org.chromium.chrome.browser.offlinepages.prefetch.PrefetchConfiguration;
 import org.chromium.chrome.browser.partnercustomizations.HomepageManager;
 import org.chromium.chrome.browser.password_manager.ManagePasswordsReferrer;
 import org.chromium.chrome.browser.password_manager.PasswordManagerLauncher;
+import org.chromium.chrome.browser.preferences.Pref;
+import org.chromium.chrome.browser.preferences.PrefServiceBridge;
 import org.chromium.chrome.browser.search_engines.TemplateUrlServiceFactory;
 import org.chromium.chrome.browser.settings.datareduction.DataReductionPreferenceFragment;
 import org.chromium.chrome.browser.settings.developer.DeveloperSettings;
@@ -54,6 +56,7 @@ public class MainSettings extends PreferenceFragmentCompat
     public static final String PREF_UI_THEME = "ui_theme";
     public static final String PREF_DATA_REDUCTION = "data_reduction";
     public static final String PREF_NOTIFICATIONS = "notifications";
+    public static final String PREF_SODIUM = "sodium";
     public static final String PREF_LANGUAGES = "languages";
     public static final String PREF_DOWNLOADS = "downloads";
     public static final String PREF_DEVELOPER = "developer";
@@ -184,6 +187,9 @@ public class MainSettings extends PreferenceFragmentCompat
     }
 
     private void updatePreferences() {
+        Preference sodium = addPreferenceIfAbsent(PREF_SODIUM);
+        setOnOffSummary(sodium, PrefServiceBridge.getInstance().getBoolean(Pref.ENABLE_SODIUM));
+
         if (IdentityServicesProvider.get().getSigninManager().isSigninSupported()) {
             addPreferenceIfAbsent(PREF_SIGN_IN);
         } else {
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/settings/SodiumSettings.java b/chrome/android/java/src/org/chromium/chrome/browser/settings/SodiumSettings.java
new file mode 100644
index 0000000..1fe9c46
--- /dev/null
+++ b/chrome/android/java/src/org/chromium/chrome/browser/settings/SodiumSettings.java
@@ -0,0 +1,49 @@
+package org.chromium.chrome.browser.settings;
+
+import android.os.Bundle;
+import android.support.v7.preference.PreferenceFragmentCompat;
+
+import org.chromium.base.ContextUtils;
+import org.chromium.chrome.R;
+import org.chromium.chrome.browser.customtabs.CustomTabActivity;
+import org.chromium.chrome.browser.preferences.Pref;
+import org.chromium.chrome.browser.preferences.PrefServiceBridge;
+import org.chromium.chrome.browser.settings.ChromeSwitchPreference;
+import org.chromium.chrome.browser.settings.SettingsUtils;
+import org.chromium.chrome.browser.settings.TextMessagePreference;
+import org.chromium.ui.text.NoUnderlineClickableSpan;
+import org.chromium.ui.text.SpanApplier;
+import org.chromium.ui.text.SpanApplier.SpanInfo;
+
+/**
+ * Fragment to manage Sodium preference and to explain to the user what it does.
+ */
+public class SodiumSettings extends PreferenceFragmentCompat {
+    private static final String PREF_SODIUM_SWITCH = "sodium_switch";
+    private static final String PREF_SODIUM_DESCRIPTION = "sodium_description";
+
+    @Override
+    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
+        SettingsUtils.addPreferencesFromResource(this, R.xml.sodium_preferences);
+        getActivity().setTitle(R.string.sodium_title);
+
+        ChromeSwitchPreference sodiumSwitch = (ChromeSwitchPreference) findPreference(PREF_SODIUM_SWITCH);
+
+        boolean isSodiumEnabled = PrefServiceBridge.getInstance().getBoolean(Pref.ENABLE_SODIUM);
+        sodiumSwitch.setChecked(isSodiumEnabled);
+
+        sodiumSwitch.setOnPreferenceChangeListener((preference, newValue) -> {
+            PrefServiceBridge.getInstance().setBoolean(Pref.ENABLE_SODIUM, (boolean) newValue);
+            return true;
+        });
+
+        TextMessagePreference sodiumDescription = (TextMessagePreference) findPreference(PREF_SODIUM_DESCRIPTION);
+        NoUnderlineClickableSpan linkSpan = new NoUnderlineClickableSpan(getResources(), view -> {
+            CustomTabActivity.showInfoPage(ContextUtils.activityFromContext(getContext()), "chrome://videomark/#/settings");
+        });
+        sodiumDescription.setSummary(
+            SpanApplier.applySpans(getString(R.string.sodium_settings_link),
+            new SpanApplier.SpanInfo("<link>", "</link>", linkSpan))
+        );
+    }
+}
diff --git a/chrome/browser/ui/android/strings/android_chrome_strings.grd b/chrome/browser/ui/android/strings/android_chrome_strings.grd
index 7284bd0..a86b6cc 100644
--- a/chrome/browser/ui/android/strings/android_chrome_strings.grd
+++ b/chrome/browser/ui/android/strings/android_chrome_strings.grd
@@ -644,6 +644,23 @@ CHAR-LIMIT guidelines:
         Darken websites
       </message>
 
+      <!-- Sodium preferences -->
+      <message name="IDS_SODIUM_TITLE" desc="Title for Sodium preference">
+        動画再生品質の記録
+      </message>
+      <message name="IDS_SODIUM_DESCRIPTION" desc="Description for Sodium preference">
+        「動画再生品質の記録」をオンにすると、YouTube、Paravi、TVer で動画を再生する際にその体感品質値の推定に必要な視聴データ (動画の解像度やビットレート、再生の一時停止など) を収集し、動画再生中や再生終了後にその推定品質値を確認できるようになります。
+      </message>
+      <message name="IDS_SODIUM_SETTINGS_LINK" desc="Link for chrome://videomark/#/settings">
+        詳細の設定については &lt;link&gt;VideoMark の設定&lt;/link&gt; をご覧ください
+      </message>
+      <message name="IDS_SODIUM_ON" desc="Accessibility description for when enable sodium.">
+        動画再生品質の記録をオフにします
+      </message>
+      <message name="IDS_SODIUM_OFF" desc="Accessibility description for when disable sodium.">
+        動画再生品質の記録をオンにします
+      </message>
+
       <!-- Privacy preferences -->
       <message name="IDS_PREFS_PRIVACY" desc="Title for the Privacy preferences. [CHAR-LIMIT=32]">
         Privacy
