diff --git a/chrome/android/chrome_java_sources.gni b/chrome/android/chrome_java_sources.gni
index a786f91..4267101 100644
--- a/chrome/android/chrome_java_sources.gni
+++ b/chrome/android/chrome_java_sources.gni
@@ -1282,6 +1282,7 @@ chrome_java_sources = [
   "java/src/org/chromium/chrome/browser/preferences/SearchEnginePreference.java",
   "java/src/org/chromium/chrome/browser/preferences/SearchUtils.java",
   "java/src/org/chromium/chrome/browser/preferences/SeekBarPreference.java",
+  "java/src/org/chromium/chrome/browser/preferences/SodiumPreference.java",
   "java/src/org/chromium/chrome/browser/preferences/SpinnerPreference.java",
   "java/src/org/chromium/chrome/browser/preferences/sync/AccountManagementFragment.java",
   "java/src/org/chromium/chrome/browser/preferences/sync/ManageSyncPreferences.java",
diff --git a/chrome/android/java/res/menu/main_menu.xml b/chrome/android/java/res/menu/main_menu.xml
index 4580a99..d444b04 100644
--- a/chrome/android/java/res/menu/main_menu.xml
+++ b/chrome/android/java/res/menu/main_menu.xml
@@ -76,6 +76,16 @@
             android:icon="@drawable/reader_mode_prefs_icon" />
         <item android:id="@+id/preferences_id"
             android:title="@string/menu_preferences" />
+        <item android:id="@+id/enable_sodium_row_menu_id"
+            android:title="@null">
+          <menu>
+              <item android:id="@+id/enable_sodium_id"
+                android:title="@string/sodium_title" />
+              <item android:id="@+id/enable_sodium_check_id"
+                android:title="@null"
+                android:checkable="true" />
+          </menu>
+        </item>
         <item android:id="@+id/help_id"
             android:title="@string/menu_help" />
         <item android:id="@+id/enter_vr_id"
diff --git a/chrome/android/java/res/xml/main_preferences.xml b/chrome/android/java/res/xml/main_preferences.xml
index a5c50e6..106dba3 100644
--- a/chrome/android/java/res/xml/main_preferences.xml
+++ b/chrome/android/java/res/xml/main_preferences.xml
@@ -71,44 +71,49 @@
         android:order="12"
         android:title="@string/prefs_section_advanced"/>
     <Preference
+        android:fragment="org.chromium.chrome.browser.preferences.SodiumPreference"
+        android:key="sodium"
+        android:order="13"
+        android:title="@string/sodium_title"/>
+    <Preference
         android:fragment="org.chromium.chrome.browser.preferences.privacy.PrivacyPreferences"
         android:key="privacy"
-        android:order="13"
+        android:order="14"
         android:title="@string/prefs_privacy"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.preferences.AccessibilityPreferences"
         android:key="accessibility"
-        android:order="14"
+        android:order="15"
         android:title="@string/prefs_accessibility"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.preferences.website.SiteSettingsPreferences"
         android:key="content_settings"
-        android:order="15"
+        android:order="16"
         android:title="@string/prefs_site_settings"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.preferences.languages.LanguagesPreferences"
         android:key="languages"
-        android:order="16"
+        android:order="17"
         android:title="@string/prefs_languages"/>
     <org.chromium.chrome.browser.preferences.ChromeBasePreference
         android:fragment="org.chromium.chrome.browser.preferences.datareduction.DataReductionPreferenceFragment"
         android:key="data_reduction"
-        android:order="17"
+        android:order="18"
         android:title="@string/data_reduction_title_lite_mode"/>
     <org.chromium.chrome.browser.preferences.ChromeBasePreference
         android:fragment="org.chromium.chrome.browser.preferences.download.DownloadPreferences"
         android:key="downloads"
-        android:order="18"
+        android:order="19"
         android:title="@string/menu_downloads"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.preferences.developer.DeveloperPreferences"
         android:key="developer"
-        android:order="19"
+        android:order="20"
         android:title="Developer options"/>
     <Preference
         android:fragment="org.chromium.chrome.browser.preferences.AboutChromePreferences"
         android:key="about_chrome"
-        android:order="20"
+        android:order="21"
         android:title="@string/prefs_about_chrome"/>
 
 </PreferenceScreen>
diff --git a/chrome/android/java/res/xml/sodium_preferences.xml b/chrome/android/java/res/xml/sodium_preferences.xml
new file mode 100644
index 0000000..09b668f
--- /dev/null
+++ b/chrome/android/java/res/xml/sodium_preferences.xml
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<PreferenceScreen
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <org.chromium.chrome.browser.preferences.ChromeSwitchPreference
+        android:key="sodium_switch"
+        android:summaryOn="@string/text_on"
+        android:summaryOff="@string/text_off" />
+
+    <org.chromium.chrome.browser.preferences.TextMessagePreference
+        android:title="@string/sodium_description"
+        app:allowDividerBelow="false" />
+
+</PreferenceScreen>
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java b/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java
index d9310b8..e386845 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/ChromeActivity.java
@@ -2190,6 +2190,10 @@ public abstract class ChromeActivity<C extends ChromeActivityComponent>
             currentTab.getWebContents().getNavigationController().setUseDesktopUserAgent(
                     !usingDesktopUserAgent, reloadOnChange);
             RecordUserAction.record("MobileMenuRequestDesktopSite");
+        } else if (id == R.id.enable_sodium_id || id == R.id.enable_sodium_check_id) {
+            final boolean reloadOnChange = !currentTab.isNativePage();
+            final boolean enabled = currentTab.getEnableSodium();
+            currentTab.setEnableSodium(!enabled, reloadOnChange);
         } else if (id == R.id.reader_mode_prefs_id) {
             DomDistillerUIUtils.openSettings(currentTab.getWebContents());
         } else {
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java b/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java
index 40fa650..cc2745d 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/appmenu/AppMenuPropertiesDelegateImpl.java
@@ -240,6 +240,7 @@ public class AppMenuPropertiesDelegateImpl implements AppMenuPropertiesDelegate
             prepareAddToHomescreenMenuItem(menu, currentTab, canShowHomeScreenMenuItem);
 
             updateRequestDesktopSiteMenuItem(menu, currentTab, true /* can show */);
+            updateEnableSodiumMenuItem(menu, currentTab, false /* temporary don't show */);
 
             // Only display reader mode settings menu option if the current page is in reader mode.
             menu.findItem(R.id.reader_mode_prefs_id)
@@ -468,6 +469,38 @@ public class AppMenuPropertiesDelegateImpl implements AppMenuPropertiesDelegate
                         : mContext.getString(R.string.menu_request_desktop_site_off));
     }
 
+    /**
+     * @param menu {@link Menu} for enable sodium.
+     * @param currentTab      Current tab being displayed.
+     */
+    protected void updateEnableSodiumMenuItem(Menu menu, Tab currentTab, boolean canShowEnableSodium) {
+        MenuItem menuRow   = menu.findItem(R.id.enable_sodium_row_menu_id);
+        MenuItem menuLabel = menu.findItem(R.id.enable_sodium_id);
+        MenuItem menuCheck = menu.findItem(R.id.enable_sodium_check_id);
+
+        // Hide enable sodium on all chrome:// pages except for the NTP.
+        String url = currentTab.getUrl();
+        boolean isChromeScheme = url.startsWith(UrlConstants.CHROME_URL_PREFIX)
+                || url.startsWith(UrlConstants.CHROME_NATIVE_URL_PREFIX);
+        // Also hide enable sodium on Reader Mode.
+        boolean isDistilledPage = DomDistillerUrlUtils.isDistilledPage(url);
+
+        boolean itemVisible = canShowEnableSodium
+                && (!isChromeScheme || currentTab.isNativePage()) && !isDistilledPage;
+        menuRow.setVisible(itemVisible);
+        if (!itemVisible) return;
+
+        boolean enabled = currentTab.getEnableSodium();
+        // Mark the checkbox if sodium is activated on this page.
+        menuCheck.setChecked(enabled);
+
+        // This title doesn't seem to be displayed by Android, but it is used to set up
+        // accessibility text in {@link AppMenuAdapter#setupMenuButton}.
+        menuLabel.setTitleCondensed(enabled
+                        ? mContext.getString(R.string.sodium_on)
+                        : mContext.getString(R.string.sodium_off));
+    }
+
     @Override
     public boolean isTranslateMenuItemVisible(Tab tab) {
         String url = tab.getUrl();
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/preferences/MainPreferences.java b/chrome/android/java/src/org/chromium/chrome/browser/preferences/MainPreferences.java
index 75983bc..92570df 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/preferences/MainPreferences.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/preferences/MainPreferences.java
@@ -53,6 +53,7 @@ public class MainPreferences extends PreferenceFragmentCompat
     public static final String PREF_UI_THEME = "ui_theme";
     public static final String PREF_DATA_REDUCTION = "data_reduction";
     public static final String PREF_NOTIFICATIONS = "notifications";
+    public static final String PREF_SODIUM = "sodium";
     public static final String PREF_LANGUAGES = "languages";
     public static final String PREF_DOWNLOADS = "downloads";
     public static final String PREF_DEVELOPER = "developer";
@@ -197,6 +198,9 @@ public class MainPreferences extends PreferenceFragmentCompat
     }
 
     private void updatePreferences() {
+        Preference sodium = addPreferenceIfAbsent(PREF_SODIUM);
+        setOnOffSummary(sodium, PrefServiceBridge.getInstance().isSodiumEnabled());
+
         if (IdentityServicesProvider.getSigninManager().isSigninSupported()) {
             addPreferenceIfAbsent(PREF_SIGN_IN);
         } else {
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/preferences/SodiumPreference.java b/chrome/android/java/src/org/chromium/chrome/browser/preferences/SodiumPreference.java
new file mode 100644
index 0000000..12fb6e5
--- /dev/null
+++ b/chrome/android/java/src/org/chromium/chrome/browser/preferences/SodiumPreference.java
@@ -0,0 +1,32 @@
+package org.chromium.chrome.browser.preferences;
+
+import android.os.Bundle;
+import android.support.v7.preference.PreferenceFragmentCompat;
+
+import org.chromium.chrome.R;
+import org.chromium.chrome.browser.preferences.ChromeSwitchPreference;
+import org.chromium.chrome.browser.preferences.PrefServiceBridge;
+import org.chromium.chrome.browser.preferences.PreferenceUtils;
+
+/**
+ * Fragment to manage Sodium preference and to explain to the user what it does.
+ */
+public class SodiumPreference extends PreferenceFragmentCompat {
+    private static final String PREF_SODIUM_SWITCH = "sodium_switch";
+
+    @Override
+    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
+        PreferenceUtils.addPreferencesFromResource(this, R.xml.sodium_preferences);
+        getActivity().setTitle(R.string.sodium_title);
+
+        ChromeSwitchPreference sodiumSwitch = (ChromeSwitchPreference) findPreference(PREF_SODIUM_SWITCH);
+
+        boolean isSodiumEnabled = PrefServiceBridge.getInstance().isSodiumEnabled();
+        sodiumSwitch.setChecked(isSodiumEnabled);
+
+        sodiumSwitch.setOnPreferenceChangeListener((preference, newValue) -> {
+            PrefServiceBridge.getInstance().setSodiumEnabled((boolean) newValue);
+            return true;
+        });
+    }
+}
diff --git a/chrome/android/java/strings/android_chrome_strings.grd b/chrome/android/java/strings/android_chrome_strings.grd
index 1206ac5..2a89e76 100644
--- a/chrome/android/java/strings/android_chrome_strings.grd
+++ b/chrome/android/java/strings/android_chrome_strings.grd
@@ -724,6 +724,20 @@ CHAR-LIMIT guidelines:
         Turn on dark theme when your device's dark theme or Battery Saver is on
       </message>
 
+      <!-- Sodium preferences -->
+      <message name="IDS_SODIUM_TITLE" desc="Title for Sodium preference">
+        動画再生品質の記録
+      </message>
+      <message name="IDS_SODIUM_DESCRIPTION" desc="Description for Sodium preference">
+        「動画再生品質の記録」をオンにすると、YouTube、Paravi、TVer で動画を再生する際にその体感品質値の推定に必要な視聴データ (動画の解像度やビットレート、再生の一時停止など) を収集し、動画再生中や再生終了後にその推定品質値を確認できるようになります。
+      </message>
+      <message name="IDS_SODIUM_ON" desc="Accessibility description for when enable sodium.">
+        動画再生品質の記録をオフにします
+      </message>
+      <message name="IDS_SODIUM_OFF" desc="Accessibility description for when disable sodium.">
+        動画再生品質の記録をオンにします
+      </message>
+
       <!-- Privacy preferences -->
       <message name="IDS_PREFS_PRIVACY" desc="Title for the Privacy preferences. [CHAR-LIMIT=32]">
         Privacy
